### PR TITLE
Allow result of PlayerInteractEvent to be changed when it's cancelled

### PR DESCRIPTION
**Pull request by williewillus**
Updated version of #3070 for 1.11.2, and adds some new test cases.
The patches look big but it's basically the same thing as #3070.

Supersedes #3435 in that that one does not account for entities.

Hope this gets merged soon.

### CODE REVIEW

File: src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java [hunk 1]
```diff
@@ -49,6 +50,7 @@
     private final BlockPos pos;
     @Nullable
     private final EnumFacing face;
+    private EnumActionResult cancellationResult = EnumActionResult.PASS;
```
Comment (liach) [inline src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java:53]: Can you keep a `final` field of the original cancellation result?

Comment (williewillus) [reply src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java:53]: Why exactly? It's always PASS.

File: patches/minecraft/net/minecraft/client/multiplayer/PlayerControllerMP.java.patch [hunk 2]
```diff
@@ -180,11 +181,12 @@
...
                  return actionresult.func_188397_a();
-@@ -500,6 +538,7 @@
+@@ -500,6 +539,8 @@
          this.func_78750_j();
          Vec3d vec3d = new Vec3d(p_187102_3_.field_72307_f.field_72450_a - p_187102_2_.field_70165_t, p_187102_3_.field_72307_f.field_72448_b - p_187102_2_.field_70163_u, p_187102_3_.field_72307_f.field_72449_c - p_187102_2_.field_70161_v);
          this.field_78774_b.func_147297_a(new CPacketUseEntity(p_187102_2_, p_187102_4_, vec3d));
-+        if(net.minecraftforge.common.ForgeHooks.onInteractEntityAt(p_187102_1_, p_187102_2_, p_187102_3_, p_187102_4_)) return EnumActionResult.PASS;
++        EnumActionResult cancelResult = net.minecraftforge.common.ForgeHooks.onInteractEntityAtAction(p_187102_1_, p_187102_2_, p_187102_3_, p_187102_4_);
++        if(cancelResult != null) return cancelResult;
          return this.field_78779_k == GameType.SPECTATOR ? EnumActionResult.PASS : p_187102_2_.func_184199_a(p_187102_1_, vec3d, p_187102_4_);
```
Comment (GotoLink) [inline patches/minecraft/net/minecraft/client/multiplayer/PlayerControllerMP.java.patch]: Could this discrepency be fixed too ? EntityInteractSpecific is called for spectators, as opposed to EntityInteract.

File: src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java [hunk 3]
```diff
@@ -208,8 +210,9 @@ public void setCanceled(boolean canceled)
     /**
      * This event is fired on both sides before the player triggers {@link net.minecraft.item.Item#onItemRightClick}.
      * Note that this is NOT fired if the player is targeting a block. For that case, see {@link RightClickBlock}.
```
Comment (GotoLink) [inline src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java]: targeting a block, or an entity...

File: src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java [hunk 4]
```diff
@@ -208,8 +210,9 @@ public void setCanceled(boolean canceled)
...
-     * Canceling the event clientside causes processing to continue to the other hands,
-     * while canceling serverside will simply do no further processing.
+     *
+     * Let result be the return value of {@link net.minecraft.item.Item#onItemRightClick}, or {@link #cancellationResult} if the event is cancelled.
+     * If we are on the client and result is not {@link EnumActionResult#SUCCESS}, the client will then continue to other hands.
```
Comment (GotoLink) [inline src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java]: I think it worth mentionning client does mainhand then, if all else fails, offhand.

Comment (williewillus) [reply src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java]: that's what it basically says there. The other events say they defer to RightClickItem on failure, and RightClickItem defers to the next hand

File: patches/minecraft/net/minecraft/client/multiplayer/PlayerControllerMP.java.patch (hunk 2 above)
Comment (williewillus) [reply patches/minecraft/net/minecraft/client/multiplayer/PlayerControllerMP.java.patch]: it makes the patch dirtier, but sure

File: src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java (hunk 3 above)
Comment (williewillus) [reply src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java]: fixing

**Comment by williewillus** [comment]
anything else I can do to speed this along?

**Comment by mezz** [comment]
I've been backlogged submitting my own PRs. 
I'll take a look at this to make sure the test mod works as intended.

**Comment by mezz** [comment]
Tested and looks good to me.
For your `Todo remove` comments, I think you should put `1.12` in there somewhere so that it's easier to find and remove in the update.

**Comment by williewillus** [comment]
Added "1.12" to the comments. Bump?

**Comment by mezz** [comment]
Still on the radar. 
The stuff merged yesterday were all very small changes.

**Pull request merged**